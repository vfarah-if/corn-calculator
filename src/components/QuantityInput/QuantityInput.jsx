import React, { useState } from "react";
import PropTypes from "prop-types";

import "./style/index.css";

const QuantityInput = ({ label, initialQuantity, onQuantityChange }) => {
  const [quantity, setQuantity] = useState(
    initialQuantity && initialQuantity >= 0 ? initialQuantity : 0
  );
  const quantityChangeHandler = (e) => {
    let value = Number.parseInt(e.target.value, 10) || 0;
    if (value < 0) {
      value = 0;
    }
    console.debug(`${label} Quantity `, value);
    setQuantity(value);
    onQuantityChange(value);
  };

  return (
    <>
      <label forhtml="quantity">{label}:</label>
      <input
        type="number"
        name="quantity"
        value={quantity}
        onChange={quantityChangeHandler}
        min={0}
      />
    </>
  );
};

QuantityInput.propTypes = {
  label: PropTypes.string.isRequired,
  initialQuantity: PropTypes.number,
  onQuantityChange: PropTypes.func.isRequired,
};

export default QuantityInput;
