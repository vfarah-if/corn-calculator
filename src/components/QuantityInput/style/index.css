label, input {
    flex-basis:200px;
  }