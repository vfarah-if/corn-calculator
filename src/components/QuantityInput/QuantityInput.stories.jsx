import React, { Fragment } from "react";
import QuantityInput from "./QuantityInput";
import { action } from "@storybook/addon-actions";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  title: "Component/QuantityInput",
  component: QuantityInput,
};

const QuantityInputTemplate = (args) => (
  <Fragment>
    <QuantityInput
      {...args}
      onQuantityChange={action("on setQuantity ...")}
    />
  </Fragment>
);

export const Default = QuantityInputTemplate.bind({});
Default.args = {
  label: "Corn Quantity",
};

export const QuantityOfOne = QuantityInputTemplate.bind({});
QuantityOfOne.args = {
  label: "Geese Quantity",
  initialQuantity: 1,
};

export const QuantityOfNegativeWillBeZero = QuantityInputTemplate.bind({});
QuantityOfNegativeWillBeZero.args = {
  label: "Quantity in the Positive",
  initialQuantity: -1,
};
