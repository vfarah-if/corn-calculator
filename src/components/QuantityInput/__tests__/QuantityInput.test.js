import React from "react";
import { fireEvent, render } from "@testing-library/react";

import {
  Default,
  QuantityOfOne,
  QuantityOfNegativeWillBeZero,
} from "../QuantityInput.stories";
import QuantityInput from "../QuantityInput";

describe("QuantityInput", () => {
  test("should create a default label and input component defaulting quantity to zero", () => {
    const { container } = render(<Default {...Default.args} />);

    expect(container).toMatchSnapshot("default");
  });

  test("should have a quantity of 1 when set or changed", () => {
    const { container } = render(<QuantityOfOne {...Default.args} />);

    expect(container).toMatchSnapshot("quantity of one");
  });

  test("should set negative initial quantity to zero", () => {
    const { container } = render(
      <QuantityOfNegativeWillBeZero {...Default.args} />
    );

    expect(container).toMatchSnapshot("negative quantity");
  });

  test("should never allow clipboard injection of a negative value", async () => {
    const expectedNumber = 0;
    const handleChange = jest.fn();
    const { container } = render(
      <QuantityInput label="Test" initialQuantity={0} onQuantityChange={handleChange} />
    );

    const inputElement = container.querySelector('input');
    fireEvent.change(inputElement, { target: { value: -1 } });

    expect(inputElement.value).toBe(expectedNumber.toString());
    expect(handleChange).toHaveBeenCalledWith(expectedNumber);
    expect(handleChange).toBeCalledTimes(1);
  });

  test("should assign a value to the onQuantityChange", async () => {
    const expectedNumber = 2;
    const handleChange = jest.fn();
    const { container } = render(
      <QuantityInput label="Test" initialQuantity={0} onQuantityChange={handleChange} />
    );

    const inputElement = container.querySelector('input');
    fireEvent.change(inputElement, { target: { value: expectedNumber } });

    expect(inputElement.value).toBe(expectedNumber.toString());
    expect(handleChange).toHaveBeenCalledWith(expectedNumber);
    expect(handleChange).toBeCalledTimes(1);
  });
});
