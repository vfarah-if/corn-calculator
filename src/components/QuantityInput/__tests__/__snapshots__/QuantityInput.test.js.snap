// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityInput should create a default label and input component defaulting quantity to zero: default 1`] = `
<div>
  <label
    forhtml="quantity"
  >
    Corn Quantity
    :
  </label>
  <input
    min="0"
    name="quantity"
    type="number"
    value="0"
  />
</div>
`;

exports[`QuantityInput should have a quantity of 1 when set or changed: quantity of one 1`] = `
<div>
  <label
    forhtml="quantity"
  >
    Corn Quantity
    :
  </label>
  <input
    min="0"
    name="quantity"
    type="number"
    value="0"
  />
</div>
`;

exports[`QuantityInput should set negative initial quantity to zero: negative quantity 1`] = `
<div>
  <label
    forhtml="quantity"
  >
    Corn Quantity
    :
  </label>
  <input
    min="0"
    name="quantity"
    type="number"
    value="0"
  />
</div>
`;
